python_onnxifi/data_conversion.cc

```diff
@@ -0,0 +1,251 @@
+    break;                                                               	 \
+  }										 \
+  case ONNX_NAMESPACE::TensorProto_DataType_UINT32:  {				 \
+    op_template<uint64_t, uint32_t>(__VA_ARGS__);                                \
+    break;                                                                       \
+  }     									 \
+  case ONNX_NAMESPACE::TensorProto_DataType_UINT64: {			 	 \
+    op_template<uint64_t, uint64_t>( __VA_ARGS__);                            	 \
+    break;							        	 \
+  }   						                        	 \
+  case ONNX_NAMESPACE::TensorProto_DataType_DOUBLE: {                   	 \
+   op_template<double, double>(__VA_ARGS__);                                 	 \
+   break;                                                                        \
+  }										 \
+  case ONNX_NAMESPACE::TensorProto_DataType_COMPLEX128: {     			 \
+   op_template<std::complex<double>, std::complex<double>>(__VA_ARGS__);         \
+   break;									 \
+  }										 \
+  case ONNX_NAMESPACE::TensorProto_DataType_STRING: 				 \
+  case ONNX_NAMESPACE::TensorProto_DataType_UNDEFINED: {			 \
+   throw std::runtime_error("Dispatch received non-numeric data type");          \
+  }										 \
+}										 \
+
+template<typename onnx_type, typename unused>
+void DataConversion::getBufferSize(uint64_t& buffer_size, const uint64_t* shape, uint32_t dimensions)  {
+  auto numElements = std::accumulate(shape, shape + dimensions,
+                                    (uint64_t) 1, std::multiplies<uint64_t>());
+  buffer_size = sizeof(onnx_type) * numElements;
+}
+
+void DataConversion::updateOutputDescriptors(ModelProto& model)  {
+  releaseDescriptors(output_descriptors_);
+  ::ONNX_NAMESPACE::shape_inference::InferShapes(model);
+  auto& outputs = model.graph().output();
+  for (auto& vi : outputs)  {
+    onnxTensorDescriptor outputDescriptor;
+    char* name = new char[vi.name().size() + 1];
+    strcpy(name, vi.name().c_str());
+    outputDescriptor.name = name;
+    outputDescriptor.memoryType = ONNXIFI_MEMORY_TYPE_CPU; //TODO: Expand memory types?
+    if (!vi.type().tensor_type().has_elem_type())  {
+      throw std::runtime_error("Non-static ModelProto: Output data type not found");
+    }
+    outputDescriptor.dataType = vi.type().tensor_type().elem_type();
+    outputDescriptor.dimensions = vi.type().tensor_type().shape().dim_size();
+    auto* shape = new uint64_t[outputDescriptor.dimensions];
+    for (auto i = 0; i < outputDescriptor.dimensions; ++i)  {
+      const auto& dim = vi.type().tensor_type().shape().dim(i);
+      if (!dim.has_dim_value())  {
+        throw std::runtime_error("Non-static ModelProto: Output shape dimension not found");
+      }
+      shape[i] = dim.dim_value();
+    }
+    outputDescriptor.shape = shape;
+    uint64_t buffer_size;
+    DISPATCH_OVER_NUMERIC_DATA_TYPE(outputDescriptor.dataType, getBufferSize, buffer_size,
```

Comment:
clang-format?

Reply:
I'm not familiar with what's wrong with the style here - could you elaborate?

Reply:
The indent seems off. Clang-format can fix it.

Reply:
using Chromimum Clang-format now